lib/Sema/TypeCheckDecl.cpp

```diff
@@ -1171,6 +1173,98 @@ doesAccessorNeedDynamicAttribute(AccessorDecl *accessor, Evaluator &evaluator) {
   }
 }
 
+llvm::Expected<bool>
+IsFinalRequest::evaluate(Evaluator &evaluator, ValueDecl *decl) const {
+  switch (decl->getKind()) {
+    case DeclKind::Var: {
+      // Properties are final if they are declared 'static' or a 'let'
+      auto *VD = cast<VarDecl>(decl);
+      if (auto *nominalDecl = VD->getDeclContext()->getSelfNominalTypeDecl()) {
+        // If this variable is a class member, mark it final if the
+        // class is final, or if it was declared with 'let'.
+        auto *PBD = VD->getParentPatternBinding();
+        if (PBD && inferFinalAndDiagnoseIfNeeded(decl, PBD->getStaticSpelling()))
+          return true;
+
+        if (VD->isLet() && isa<ClassDecl>(nominalDecl)) {
+          if (VD->getFormalAccess() == AccessLevel::Open) {
+            auto &context = decl->getASTContext();
+            auto diagID = diag::implicitly_final_cannot_be_open;
+            if (!context.isSwiftVersionAtLeast(5))
+              diagID = diag::implicitly_final_cannot_be_open_swift4;
+            auto inFlightDiag =
+              context.Diags.diagnose(decl, diagID,
+                                     static_cast<unsigned>(ImplicitlyFinalReason::Let));
+            fixItAccess(inFlightDiag, decl, AccessLevel::Public);
+          }
+
+          return true;
+        }
+      }
+
+      break;
+    }
+
+    case DeclKind::Func:
+      if (decl->getDeclContext()->isTypeContext()) {
+        // Method declared 'static' are final.
+        auto staticSpelling = cast<FuncDecl>(decl)->getStaticSpelling();
+        if (inferFinalAndDiagnoseIfNeeded(decl, staticSpelling))
+          return true;
+      }
+      break;
+
+    case DeclKind::Accessor:
+      if (auto accessor = dyn_cast<AccessorDecl>(decl)) {
+        switch (accessor->getAccessorKind()) {
+          case AccessorKind::DidSet:
+          case AccessorKind::WillSet:
+            // Observing accessors are marked final if in a class.
+            if (accessor->getDeclContext()->getSelfClassDecl())
+              return true;
+            break;
+
+          case AccessorKind::Read:
+          case AccessorKind::Modify: {
+            // Coroutines are final if their storage is.
+            auto storage = accessor->getStorage();
+            if (storage->isFinal())
+              return true;
+            break;
+          }
+
+          default:
+            break;
+        }
+
+        if (accessor->isGetterOrSetter()) {
```

Comment:
This is the same as the coroutine case. It would be clearer if you just folded it into the above switch statement.